libs/ui/src/reboot_from_usb_button.tsx

```diff
@@ -0,0 +1,147 @@
+  const attemptReboot = useCallback(async () => {
+    assert(window.kiosk);
+    await logger.log(LogEventId.PrepareBootFromUsbInit, 'superadmin', {
+      message:
+        'Attempting to set the USB drive next in the boot order and reboot',
+    });
+    if (usbDriveStatus !== usbstick.UsbDriveStatus.mounted) {
+      setCurrentState(State.NO_USB_INSERTED);
+      await logger.log(LogEventId.PrepareBootFromUsbComplete, 'superadmin', {
+        message: 'No USB drive found, could not boot from usb',
+        disposition: 'failure',
+      });
+      return;
+    }
+    setCurrentState(State.PREPARING_BOOT);
+    const readyToReboot = await window.kiosk.prepareToBootFromUsb();
+    if (readyToReboot) {
+      await logger.log(LogEventId.PrepareBootFromUsbComplete, 'superadmin', {
+        message:
+          'USB Drive successfully set next in the boot order for the next boot.',
+        disposition: 'success',
+      });
+      await logger.log(LogEventId.RebootMachine, 'superadmin', {
+        message: 'Machine will now reboot…',
+      });
+      setCurrentState(State.REBOOTING);
+      await window.kiosk.reboot();
+    } else {
+      setCurrentState(State.NO_BOOTABLE_USB_OPTION);
+      await logger.log(LogEventId.PrepareBootFromUsbComplete, 'superadmin', {
+        message:
+          'USB Drive not found in the list of bootable options, boot order is not modified.',
+        disposition: 'failure',
+      });
+    }
+  }, [usbDriveStatus, logger]);
+  async function reboot() {
+    assert(window.kiosk);
+    await logger.log(LogEventId.RebootMachine, 'superadmin', {
+      message: 'User trigged a reboot of the machine…',
+    });
+    setCurrentState(State.REBOOTING);
+    await window.kiosk.reboot();
+  }
+  useEffect(() => {
+    if (
+      usbDriveStatus === usbstick.UsbDriveStatus.mounted &&
+      currentState === State.NO_USB_INSERTED
+    ) {
+      void attemptReboot();
+    }
+  }, [usbDriveStatus, currentState, attemptReboot]);
+
+  const onClose = useCallback(() => setCurrentState(State.CLOSED), [
+    setCurrentState,
+  ]);
+
+  if (currentState === State.PREPARING_BOOT) {
+    return <Modal content={<Loading>Preparing to boot…</Loading>} />;
+  }
+  if (currentState === State.REBOOTING) {
+    return <Modal content={<Loading>Rebooting…</Loading>} />;
+  }
+  if (currentState === State.NO_BOOTABLE_USB_OPTION) {
+    return (
+      <Modal
+        content={
+          <Prose>
+            The USB Drive was not found in the list of bootable devices. Make
+            sure it is the correct USB drive and reboot the machine with it
+            plugged in, then try again.
+          </Prose>
+        }
+        onOverlayClick={onClose}
+        actions={
+          <React.Fragment>
+            <Button primary onPress={reboot}>
+              Reboot
+            </Button>
+            <Button onPress={onClose}>Close</Button>
+          </React.Fragment>
+        }
+      />
+    );
+  }
```

Comment:
This case doesn't make sense to me. Is it trying to say that the currently inserted USB drive is not bootable? What will pressing "Reboot" do? Just reboot back into the app? If so, why even allow rebooting here? I think this needs to be clarified.

Reply:
It depends on the hardware but most of them only update the devices that `efibootmgr` will list as bootable when the machine boots. So if you insert a bootable usb stick after the machine has booted you actually need to reboot once before it will work.